CHANGELOG.rst

```diff
@@ -5,49 +5,48 @@ Changelog
 TBR
 ---
 
-* Scrapy uses the spider's ``parse()`` method when the :class:`scrapy.http.Request`
-  callback is set to ``None``.
+* Improved ``scrapy-poet`` behavior when ``spider.parse()`` method
+  arguments are annotated. Previously, they were causing unnecessary work
+  in unexpected places like
+  :class:`scrapy.downloadermiddlewares.robotstxt.RobotsTxtMiddleware`,
+  :class:`scrapy.pipelines.images.ImagesPipeline` or
+  :class:`scrapy.pipelines.files.FilesPipeline`. It is also a reason
+  ``HttpClient`` might not be working in page objects.
+  Now these cases are detected, and a warning is issued.
+
+  As of Scrapy 2.7, it is not possible to fix the issue completely
+  in scrapy-poet. Fixing it would require Scrapy changes; some 3rd party
+  libraries may also need to be updated.
+
+  .. note::
+
+      The root of the issue is that when request.callback is None, ``parse``
+      callback is assumed normally. But sometimes callback=None is used
+      when :class:`scrapy.http.Request` is added to the Scrapy's downloader
+      directly, in which case no callback is used. Middlewares, including
+      scrapy-poet's, can't distinguish between these two cases, which causes
+      all kinds of issues.
+
+  We recommend all ``scrapy-poet`` users to modify their code to
+  avoid the issue. Please **don't** define ``parse``
+  method with arguments which are supposed to be filled by ``scrapy-poet``,
```

Comment:
The main recommendation in the [Pitfalls](https://scrapy-poet.readthedocs.io/en/latest/intro/pitfalls.html) doc is to simply avoid defining the `parse()` method and choose any other name.

I think this is much simpler than a user checking if their `parse()` method has some dependencies fulfilled by **scrapy-poet**.

I also think it's less error prone in the future when users forget that this issue exists and they'd be adding dependencies to it. If they'd renamed the `parse()` method from the start, they'd be immune to future issues.

Reply:
> I think this is much simpler than a user checking if their parse() method has some dependencies fulfilled by scrapy-poet.

Checking if a dependency is used looks simple enough: you find parse method, and see if it has arguments other than response. 

> I also think it's less error prone in the future when users forget that this issue exists and they'd be adding dependencies to it. If they'd renamed the parse() method from the start, they'd be immune to future issues.

I think the current workaround won't be there forever; we'll make a fix in Scrapy (hopefully, pretty soon) to avoid this issue, so that users will be able to use parse methods again. It also looks like if they add dependencies to parse method, a warning is going to be issued, and it won't work anyways, so hopefully the error would be obvious.

Is it ok with you to keep the recommendation of changing parse method only if it uses scrapy-poet provided arguments, at least in the changelog, to minimize the suggested changes users need to do to upgrade?

Reply:
> Is it ok with you to keep the recommendation of changing parse method only if it uses scrapy-poet provided arguments, at least in the changelog, to minimize the suggested changes users need to do to upgrade?

Yeah, let's go with it in the changelog. The docs provide more information anyways which could also help them determine the best course of action in their use case. :)